### PR TITLE
Refcount responses in `TransportNodesAction`

### PR DESCRIPTION
**Pull request by DaveCTurner**
Today we `decRef()` the per-node responses just after adding them to the
`responses` collection, but in fact we should keep them alive until
we've constructed the final response. This commit does that.

### CODE REVIEW

**Comment by DaveCTurner** [comment]
Ah hmm actually this will leak the responses on cancellation won't it? Sorry, needs more work here.

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java [hunk 1]
```diff
@@ -124,6 +126,23 @@ protected void doExecute(Task task, NodesRequest request, ActionListener<NodesRe
 
             final TransportRequestOptions transportRequestOptions = TransportRequestOptions.timeout(request.timeout());
 
+            {
+                addReleaseOnCancellationListener();
+            }
+
+            private void addReleaseOnCancellationListener() {
```
Comment (DaveCTurner) [inline server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:133]: Incredibly if you write this inline in the anonymous constructor then the compiler blows up with a NPE: https://gradle-enterprise.elastic.co/s/pp3aym5qktpa2

**Review by fcofdez — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java [hunk 2]
```diff
@@ -96,6 +98,23 @@ protected void doExecute(Task task, NodesRequest request, ActionListener<NodesRe
...
+                if (task instanceof CancellableTask cancellableTask) {
+                    cancellableTask.addListener(() -> {
+                        final List<NodeResponse> drainedResponses;
+                        synchronized (responses) {
+                            drainedResponses = List.copyOf(responses);
```
Comment (arteam) [inline server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:110]: I was wondering if could add copy and remove elements in one iteration:

```
var iterator = responses.iterator();
while (iterator.hasNext()) {
    drainedResponses.add(iterator.next());
    iterator.remove();
}
```

**Review by arteam — APPROVED** [review]
LGTM

File: server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java (hunk 2 above)
Comment (DaveCTurner) [reply server/src/main/java/org/elasticsearch/action/support/nodes/TransportNodesAction.java:110]: You can, but adding elements one-by-one to a list potentially involves more allocations to grow the list incrementally, and removing items from a list iterator one-by-one is an O(N²) operation.

**Comment by DaveCTurner** [comment]
`elasticsearch-ci/docs` isn't working right now but it duplicates `buildkite/docs-build-pr` which has passed so I'm merging this.

**Pull request merged**